File: string/24_palindrome_pairs.py
"""
    Palindrome Pairs - LeetCode #336
    
    Given a list of unique words, return all the pairs of distinct indices (i, j) in the
    given list, so that the concatenation of the two words words[i] + words[j] is a palindrome.
    
    Example 1:
    Input: words = ["abcd","dcba","lls","s","sssll"]
    Output: [[0,1],[1,0],[3,2],[2,4]]
    Explanation: The palindromes are ["dcbaabcd","abcddcba","slls","llssssll"]
    
    Example 2:
    Input: words = ["bat","tab","cat"]
    Output: [[0,1],[1,0]]
    Explanation: The palindromes are ["battab","tabbat"]
    
    Example 3:
    Input: words = ["a",""]
    Output: [[0,1],[1,0]]
    
    Constraints:
    - 1 <= words.length <= 5000
    - 0 <= words[i].length <= 300
    - words[i] consists of lowercase English letters
    
    Time Complexity: O(n * k^2) where k is average word length
    Space Complexity: O(n)
"""

from typing import List

class Solution:
    def palindromePairs(self, words: List[str]) -> List[List[int]]:
        result = []
        word_map = {word: i for i, word in enumerate(words)}
        
        for i, word in enumerate(words):
            length = len(word)
            
            # Check all possible splits
            for j in range(length + 1):
                left = word[:j]
                right = word[j:]
                
                # If left is palindrome, check if reverse of right exists
                if self.is_palindrome(left):
                    reverse_right = right[::-1]
                    if reverse_right in word_map and word_map[reverse_right] != i:
                        result.append([word_map[reverse_right], i])
                
                # If right is palindrome, check if reverse of left exists
                # j != length to avoid duplicates
                if j != length and self.is_palindrome(right):
                    reverse_left = left[::-1]
                    if reverse_left in word_map and word_map[reverse_left] != i:
                        result.append([i, word_map[reverse_left]])
        
        return result
    
    def is_palindrome(self, s: str) -> bool:
        return s == s[::-1]

if __name__ == "__main__":
    sol = Solution()
    
    # Test 1
    words1 = ["abcd", "dcba", "lls", "s", "sssll"]
    result1 = sol.palindromePairs(words1)
    print(f"Test 1: {result1}")
    
    # Test 2
    words2 = ["bat", "tab", "cat"]
    result2 = sol.palindromePairs(words2)
    print(f"Test 2: {result2}")

File: string/24_palindrome_pairs.cpp
/*
    Palindrome Pairs - LeetCode #336
    
    Given a list of unique words, return all the pairs of distinct indices (i, j) in the
    given list, so that the concatenation of the two words words[i] + words[j] is a palindrome.
    
    Example 1:
    Input: words = ["abcd","dcba","lls","s","sssll"]
    Output: [[0,1],[1,0],[3,2],[2,4]]
    Explanation: The palindromes are ["dcbaabcd","abcddcba","slls","llssssll"]
    
    Example 2:
    Input: words = ["bat","tab","cat"]
    Output: [[0,1],[1,0]]
    Explanation: The palindromes are ["battab","tabbat"]
    
    Example 3:
    Input: words = ["a",""]
    Output: [[0,1],[1,0]]
    
    Constraints:
    - 1 <= words.length <= 5000
    - 0 <= words[i].length <= 300
    - words[i] consists of lowercase English letters
    
    Time Complexity: O(n * k^2) where k is average word length
    Space Complexity: O(n)
*/

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> result;
        unordered_map<string, int> wordMap;
        
        // Build word to index map
        for (int i = 0; i < words.size(); i++) {
            wordMap[words[i]] = i;
        }
        
        for (int i = 0; i < words.size(); i++) {
            string word = words[i];
            int len = word.length();
            
            // Check all possible splits
            for (int j = 0; j <= len; j++) {
                string left = word.substr(0, j);
                string right = word.substr(j);
                
                // If left is palindrome, check if reverse of right exists
                if (isPalindrome(left)) {
                    string reverseRight = right;
                    reverse(reverseRight.begin(), reverseRight.end());
                    
                    if (wordMap.count(reverseRight) && wordMap[reverseRight] != i) {
                        result.push_back({wordMap[reverseRight], i});
                    }
                }
                
                // If right is palindrome, check if reverse of left exists
                // j != len to avoid duplicates
                if (j != len && isPalindrome(right)) {
                    string reverseLeft = left;
                    reverse(reverseLeft.begin(), reverseLeft.end());
                    
                    if (wordMap.count(reverseLeft) && wordMap[reverseLeft] != i) {
                        result.push_back({i, wordMap[reverseLeft]});
                    }
                }
            }
        }
        
        return result;
    }
    
private:
    bool isPalindrome(const string& s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s[left++] != s[right--]) return false;
        }
        return true;
    }
};

int main() {
    Solution sol;
    
    // Test 1
    vector<string> words1 = {"abcd", "dcba", "lls", "s", "sssll"};
    vector<vector<int>> result1 = sol.palindromePairs(words1);
    cout << "Test 1: ";
    for (const auto& pair : result1) {
        cout << "[" << pair[0] << "," << pair[1] << "] ";
    }
    cout << endl;
    
    // Test 2
    vector<string> words2 = {"bat", "tab", "cat"};
    vector<vector<int>> result2 = sol.palindromePairs(words2);
    cout << "Test 2: ";
    for (const auto& pair : result2) {
        cout << "[" << pair[0] << "," << pair[1] << "] ";
    }
    cout << endl;
    
    return 0;
}
